Objects/structseq.c

```diff
@@ -287,78 +287,48 @@ static PyMethodDef structseq_methods[] = {
-    0,                                          /* tp_descr_get */
-    0,                                          /* tp_descr_set */
-    0,                                          /* tp_dictoffset */
-    0,                                          /* tp_init */
-    0,                                          /* tp_alloc */
-    structseq_new,                              /* tp_new */
-};
+static void count_members(PyStructSequence_Desc *desc, Py_ssize_t *n_members, Py_ssize_t *n_unnamed_members) {
+    Py_ssize_t i;
 
-int
-PyStructSequence_InitType2(PyTypeObject *type, PyStructSequence_Desc *desc)
-{
-    PyObject *dict;
-    PyMemberDef* members;
-    Py_ssize_t n_members, n_unnamed_members, i, k;
+    *n_unnamed_members = 0;
+    for (i = 0; desc->fields[i].name != NULL; ++i)
+        if (desc->fields[i].name == PyStructSequence_UnnamedField)
+            (*n_unnamed_members)++;
+        (*n_members) = i;
+}
+
+static int initialize_structseq_dict(PyStructSequence_Desc *desc, PyObject* dict) {
     PyObject *v;
+    Py_ssize_t n_members, n_unnamed_members;
 
-#ifdef Py_TRACE_REFS
-    /* if the type object was chained, unchain it first
-       before overwriting its storage */
-    if (type->ob_base.ob_base._ob_next) {
-        _Py_ForgetReference((PyObject*)type);
-    }
-#endif
+#define SET_DICT_FROM_SIZE(key, value)                          \
+    do {                                                        \
+        v = PyLong_FromSsize_t(value);                          \
+        if (v == NULL)                                          \
+            return -1;                                          \
+        if (PyDict_SetItemString(dict, key, v) < 0) {           \
+            Py_DECREF(v);                                       \
+            return -1;                                          \
+        }                                                       \
+        Py_DECREF(v);                                           \
+    } while (0)
 
-    n_unnamed_members = 0;
-    for (i = 0; desc->fields[i].name != NULL; ++i)
-        if (desc->fields[i].name == PyStructSequence_UnnamedField)
-            n_unnamed_members++;
-    n_members = i;
+    count_members(desc, &n_members, &n_unnamed_members);
+    SET_DICT_FROM_SIZE(visible_length_key, desc->n_in_sequence);
+    SET_DICT_FROM_SIZE(real_length_key, n_members);
+    SET_DICT_FROM_SIZE(unnamed_fields_key, n_unnamed_members);
+    return 0;
+}
 
-    memcpy(type, &_struct_sequence_template, sizeof(PyTypeObject));
-    type->tp_base = &PyTuple_Type;
-    type->tp_name = desc->name;
-    type->tp_doc = desc->doc;
+static PyMemberDef* initialize_members(PyStructSequence_Desc *desc) {
+    PyMemberDef* members;
+    Py_ssize_t n_members, n_unnamed_members, i, k;
 
-    members = PyMem_NEW(PyMemberDef, n_members-n_unnamed_members+1);
+    count_members(desc, &n_members, &n_unnamed_members);
+    members = PyMem_NEW(PyMemberDef, n_members-n_unnamed_members + 1);
```

Comment:
I wonder if this kind of bug shows up anywhere else in Python.  I.e. the various sub-structs used by type objects are heap allocated but then don't get freed when the type gets deallocated.  Another solution (rather than writing a special tp_dealloc method would perhaps be to add a new type flag.  Something to say these sub-structs are heap allocated and when typeobject.c frees the type struct, it should also free these sub-structs.

Reply:
This leak has now been fixed in the lastest commit.